torch/nested/_internal/offload_tensor.py

```diff
@@ -0,0 +1,225 @@
+    def __init__(self, offload_hook: Optional[Callable] = None):
+        self.wrapper_refs = []
+        # Upon execution, the hook
+        # hook(host_tensor, device_tensor) -> None
+        self.offload_hook = None
+
+    def create(self, device, device_tensor, host_tensor):
+        ret = OffloadTensor(
+            device=device,
+            device_tensor=device_tensor,
+            host_tensor=host_tensor,
+            offload_hook=self.offload_hook,
+        )
+        self.wrapper_refs.append(weakref.ref(ret))
+        return ret
+
+    def maybe_offload_all(self):
+        global _global_is_pending_offload
+
+        if _global_is_pending_offload:
+            still_alive = []
+            for ref in self.wrapper_refs:
+                if (t := ref()) is not None:
+                    t.offload()
+                    still_alive.append(weakref.ref(t))
+            self.wrapper_refs = still_alive
+            _global_is_pending_offload = False
+        return
+
+    def copy(self):
+        # Is this what we want to do?
+        ret = OffloadTensorRegistry()
+        ret.wrapper_refs = self.wrapper_refs.copy()
+
+
+_global_offload_tensor_registry = None
+
+
+def init_offload_tensor_registry(registry):
+    global _global_offload_tensor_registry
+    _global_offload_tensor_registry = registry
+
+
+@contextmanager
+def offload_tensor_registry(registry):
+    global _global_offload_tensor_registry
+    # Save the current registry to restore it later
+    old_registry = _global_offload_tensor_registry
+    # Set the new registry
+    _global_offload_tensor_registry = registry
+    try:
+        yield
+    finally:
+        # Restore the old registry
+        _global_offload_tensor_registry = old_registry
+
+
+def create_offload_tensor(device, device_tensor, host_tensor):
+    assert _global_offload_tensor_registry is not None
+    if _try_get_fake_mode(device_tensor if device_tensor is not None else host_tensor):
+        return OffloadTensor(
+            device=device,
+            device_tensor=device_tensor,
+            host_tensor=host_tensor,
+        )
+    else:
+        return _global_offload_tensor_registry.create(
+            device, device_tensor, host_tensor
+        )
+
+
+def maybe_offload_all(device_tensor):
+    assert _global_offload_tensor_registry is not None
+    if _try_get_fake_mode(device_tensor):
+        # Do nothing here. Offloading is an eager-only optimization.
+        return
+    else:
+        _global_offload_tensor_registry.maybe_offload_all()
+
+
+# Offload tensor wrapper is disposable
+def _make_offload_tensor(
+    device_tensor: Optional[torch.Tensor] = None,
+    host_tensor: Optional[torch.Tensor] = None,
+    device: Optional[torch.device] = None,
+) -> "OffloadTensor":
+    assert device_tensor is not None or host_tensor is not None
+    assert device is not None or device_tensor is not None
+    maybe_offload_all(device_tensor if device_tensor is not None else host_tensor)
```

Comment:
discussed elsewhere, but for posterity: I think the semantics of possibly offloading previous upon new OffloadTensor construction are unexpected and have potentially bad implications (i.e. never offloading even after an offload request because a new OffloadTensor is never constructed)